spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/lock/JdbcLockRegistry.java

```diff
@@ -0,0 +1,232 @@
+		String path = pathFor((String) lockKey);
+		JdbcLock lock = this.locks.get(path);
+		if (lock == null) {
+			synchronized (this.locks) {
+				lock = this.locks.get(path);
+				if (lock == null) {
+					lock = new JdbcLock(this.client, path);
+					this.locks.put(path, lock);
+				}
+			}
+		}
+		return lock;
+	}
+
+	private String pathFor(String lockKey) {
+		MessageDigest digest;
+		try {
+			digest = MessageDigest.getInstance("MD5");
+		}
+		catch (NoSuchAlgorithmException e) {
+			throw new IllegalStateException("MD5 algorithm not available.  Fatal (should be in the JDK).");
+		}
+		try {
+			byte[] bytes = digest.digest(lockKey.getBytes("UTF-8"));
+			return String.format("%032x", new BigInteger(1, bytes));
+		}
+		catch (UnsupportedEncodingException e) {
+			throw new IllegalStateException("UTF-8 encoding not available.  Fatal (should be in the JDK).");
+		}
+	}
+
+	@Override
+	public void expireUnusedOlderThan(long age) {
+		synchronized (this.locks) {
+			Iterator<Entry<String, JdbcLock>> iterator = this.locks.entrySet().iterator();
+			long now = System.currentTimeMillis();
+			while (iterator.hasNext()) {
+				Entry<String, JdbcLock> entry = iterator.next();
+				JdbcLock lock = entry.getValue();
+				if (now - lock.getLastUsed() > age && !lock.isAcquiredInThisProcess()) {
+					iterator.remove();
+				}
+			}
+		}
+	}
+
+	private static final class JdbcLock implements Lock {
+
+		private final JdbcClient mutex;
+
+		private final String path;
+
+		private volatile long lastUsed = System.currentTimeMillis();
+
+		private ReentrantLock delegate = new ReentrantLock();
+
+		private JdbcLock(JdbcClient client, String path) {
+			this.mutex = client;
+			this.path = path;
+		}
+
+		public long getLastUsed() {
+			return this.lastUsed;
+		}
+
+		@Override
+		public void lock() {
+			this.delegate.lock();
+			try {
+				boolean acquired = this.mutex.acquire(this.path);
+				if (acquired) {
+					this.lastUsed = System.currentTimeMillis();
+				}
+			}
+			catch (Exception e) {
+				this.delegate.unlock();
+				throw new RuntimeException("Failed to aquire mutex at " + this.path, e);
+			}
+		}
+
+		@Override
+		public void lockInterruptibly() throws InterruptedException {
+			boolean locked = false;
+			if (this.delegate.isHeldByCurrentThread()) {
+				locked = true;
+			}
+			this.delegate.lockInterruptibly();
+			if (locked) {
+				return;
+			}
+			// this is a bit ugly, but...
+			while (!locked) {
+				try {
+					locked = doLock();
+				}
+				catch (Exception e) {
+					this.delegate.unlock();
+					throw e;
+				}
+			}
+		}
+
+		@Override
+		public boolean tryLock() {
+			try {
+				return tryLock(0, TimeUnit.MICROSECONDS);
+			}
+			catch (InterruptedException e) {
+				Thread.currentThread().interrupt();
+				return false;
+			}
+		}
+
+		@Override
+		public boolean tryLock(long time, TimeUnit unit) throws InterruptedException {
+			if (!this.delegate.tryLock(time, unit)) {
+				return false;
+			}
+			try {
+				return doLock();
```

Comment:
`RedisLockRegistry` has something like this:
```
try {
	long expire = System.currentTimeMillis() + TimeUnit.MILLISECONDS.convert(time, unit);
	boolean acquired = false;
	while (!(acquired = this.obtainLock()) && System.currentTimeMillis() < expire) {
		Thread.sleep(100);
	}
	if (!acquired) {
		localLock.unlock();
	}
		return acquired;
}
```

I don't say that you do something wrong, I just don't feel the proper solution and try to share as much mind storm as possible and we may fix it together with some middle ground.

Sorry if I sound like ignorant.